### ISSUE TITLE
Potential conflict between Optuna prunners and early stopping conditions

### ISSUE THREAD

**Issue by pedroig** [optuna/optuna#2506]
While training a model, I would like to prevent overfitting by setting an early stop condition so if I have more than `n` consecutive steps in which my metric doesn't improve, I would stop the training. Therefore, doing hyperparameter search with Optuna, this stopping criterion I described would only look at previous steps within the current trial. However, could I use this when using Optuna pruning strategies without having issues? 

To illustrate the doubt, let's say I'm following [this example we have for cv with xgboost](https://github.com/optuna/optuna/blob/9f89399ecef524f0a7b51b1c7697c22a24909060/examples/xgboost/xgboost_cv_integration.py#L19) and I change only [this line](https://github.com/optuna/optuna/blob/9f89399ecef524f0a7b51b1c7697c22a24909060/examples/xgboost/xgboost_cv_integration.py#L44) to add the `early_stopping_rounds` parameter, resulting in the following code:

```python
history = xgb.cv(param, dtrain, num_boost_round=100, callbacks=[pruning_callback], early_stopping_rounds=10)
```

This would mean that not all successful trials would have the same number of steps which maybe could harm the `MedianPruner` strategy. How would the pruning work if I have a trial on a step that was not reached by all my intermediate results of previous trials? Would it simply not be pruned? Also, what if I have previous trials for that step but the number of trials that reached it is smaller than `n_startup_trials`? Would the pruning be ignored until we have enough trials that reached that step or would it be applied as long as the total current number of trials is bigger than `n_startup_trials`?

**Comment by nzw0301** [comment]
Hi, @pedroig, I try to answer your questions because recently I looked at the implementation of `MedianPruner` and its parent class `PercentilePruner` by chance and answered a similar issue https://github.com/optuna/optuna/issues/2441.

##  Q1. if I have a trial on a step that was not reached by all my intermediate results of previous trials? Would it simply not be pruned?

You are right. Such a trial will not be pruned. We can find it by checking 
`_get_percentile_intermediate_result_over_trials` function in [the implementation](https://github.com/optuna/optuna/blob/master/optuna/pruners/_percentile.py).

## Q2. What if I have previous trials for that step but the number of trials that reached it is smaller than `n_startup_trials`?

Both `MedianPruner` and `PercentilePruner` can prune such a trial based on fewer trials than `n_startup_trials`. `n_startup_trials` only cares about the number of **completed trials** and doesn't care about the number of `steps`.

---

I proposed a new argument not to perform pruning if the number of reported steps is not enough as in https://github.com/optuna/optuna/issues/2441#issuecomment-792246044. If you have any comments on the feature, please let me know!

**Comment by pedroig** [comment]
Hi @nzw0301, thanks for the fast response! Your feature suggestion sounds great to me. However, I have a small preference for changing the behavior only via `n_startup_trials` which is a bit more restrictive but prevents other users from going through the same problem by mistake. It would also avoid the corner case of skipping pruning when there are no previous trials that reached a given step. I got your point about being backward compatible but I would say it makes sense to treat this as a minor bug and not as a feature request. Then again, this is just my personal preference, and both solutions would work equally well as long as the user is aware of the details.

**Comment by nzw0301** [comment]
Hi @pedroig, the feature has been merged into the master branch of optuna by adding the new argument rather than fixing the behaviour of pruners due to backward compatibility. The next stable release will be within one month!
